src/main/scala/com/fulcrumgenomics/umi/UmiConsensusCaller.scala

```diff
@@ -0,0 +1,277 @@
+
+import scala.collection.JavaConversions.iterableAsScalaIterable
+import scala.collection.mutable
+
+/**
+  * Contains shared types and functions used when writing UMI-driven consensus
+  * callers that take in SAMRecords and emit SAMRecords.
+  */
+object UmiConsensusCaller {
+  /** The type of consensus read to output. */
+  object ReadType extends Enumeration {
+    type ReadType = Value
+    val Fragment, FirstOfPair, SecondOfPair = Value
+  }
+
+  /**
+    * Stores information about a read to be fed into a consensus.
+    */
+  case class SourceRead(id: String, bases: Array[Byte], quals: Array[Byte]) {
+    require(bases.length == quals.length, "Bases and qualities are not the same length.")
+    def length: Int = bases.length
+  }
+
+  /** A trait that consensus reads must implement. */
+  trait ConsensusRead {
+    /** The ID of the molecule that generated the consensus read. */
+    def id: String
+    /** The bases of the consensus read. */
+    def bases: Array[Byte]
+    /** The quals of the consensus read. */
+    def quals: Array[Byte]
+    /** Gets the length of the consensus read. */
+    def length: Int = bases.length
+    /** Returns the consensus read a String - mostly useful for testing. */
+    def baseString = new String(bases)
+  }
+
+  /**
+    * Attempts to construct a String that can be used as a prefix for consensus read names based
+    * on the contents of the incoming SAMFileHeader.
+    */
+  def makePrefixFromSamHeader(header: SAMFileHeader): String = {
+    val ids = header.getReadGroups.map(rg => Option(rg.getLibrary).getOrElse(rg.getReadGroupId)).toList.sorted.distinct
+    // Read names have to fit into 255 bytes, and this is just the prefix
+    if (ids.map(_.length+1).sum <= 200) ids.mkString("|")
+    else Integer.toHexString(new Murmur3(1).hashUnencodedChars(ids.mkString("|")))
+  }
+
+  /**
+    * Constructs an output header with a single read group for the a BAM.
+    */
+  def outputHeader(in: SAMFileHeader, readGroupId: String, sortOrder: Option[SortOrder] = None): SAMFileHeader = {
+    val oldRgs = in.getReadGroups.toSeq
+    def collapse(f: SAMReadGroupRecord => String): String = oldRgs.map(f).filter(_ != null).toSet.toList match {
+      case Nil      => null
+      case x :: Nil => x
+      case xs       => xs.mkString(",")
+    }
+
+    val rg = new SAMReadGroupRecord(readGroupId)
+    rg.setDescription(s"Consensus reads generated from ${oldRgs.size} input read groups.")
+    rg.setLibrary         (collapse(_.getLibrary))
+    rg.setSample          (collapse(_.getSample))
+    rg.setPlatform        (collapse(_.getPlatform))
+    rg.setPlatformUnit    (collapse(_.getPlatformUnit))
+    rg.setSequencingCenter(collapse(_.getSequencingCenter))
+
+    val outHeader = new SAMFileHeader
+    outHeader.addReadGroup(rg)
+    outHeader.setSortOrder(sortOrder.getOrElse(SortOrder.unsorted))
+    outHeader.setGroupOrder(GroupOrder.query)
+    outHeader
+  }
+}
+
+
```

Comment:
I _always_ leave two blank lines between top level classes/traits/objects in the same file as I find the visual break useful!